# Анализ баз данных для компании, приобретшей сервис для чтения книг по подписке

<div style="border:solid orange 2px; padding: 20px"> <h7 style="color:black; margin-bottom:20px">Привет! Меня зовут Егошин Михаил, я буду проводить ревью на этом спринте. Если ты не против, давай общаться на "ты", так я думаю будет комфортнее и удобнее)
Мои комментарии и замечания далее по тексту помечены различными цветами:
    
---    
<div class="alert alert-block alert-danger">
Критическое замечание, которые следует исправить.
</div> 
    
---
    
<div class="alert alert-block alert-warning">
Замечание\совет на будущее.
</div>
    
---
    
<div class="alert alert-block alert-success">
Когда всё сделано правильно.
</div>
    
---

Для меня главное, не просто указать на ошибки, а чтобы у тебя, в первую очередь, было понимание как это делается, подсказать в каких то моментах. Поэтому если возникнут вопросы по проекту или вопросы общего характера, то не стесняйся - задавай!) Для того, чтобы мне было легче их найти, то выделяй их пожалуйста с помощью цвета, например вот так - <font color=orange>"вопрос, который не останется незамеченным"</font>
    
Удачи!
    
p.s.: не удаляй мои замечания, если предстоит что то доработать в проекте.</h7>

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Все замечания учтены. Успехов в дальнейшем обучении!
</div>

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

Итог ревью - ты проделал большую работу. Запросы понятны и логичны, но не хватает совсем чуть чуть. Есть задания, которые тебе предстоит доработать, чтобы я смог принять твой проект. Если возникнут вопросы по моим комментариям, пиши здесь - обязательно отвечу. Буду ждать!
</div>

Получение данных, которые помогут сформулировать ценностное предложение для нового продукта.

**Для этого необходимо:**
1. *Подключиться к базе SQL.*

2. *Изучить таблицы*

3.  *Выполнить поставленные заказчиком задания.*

4.  *Сопроводить выполненные задания краткими выводами и возможными комментариями.*

**Описание данных:**

**Таблица books**

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства.

**Таблица ratings**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

<a id='contents'></a>
**Содержание:**

- Шаг 1 - [Изучение общей информации о данных](#data_import)
- Шаг 2 - [Анализ данных](#data_analysis)
- [Общий вывод](#general_conclusion)

*Импортируем библиотеки:*

In [1]:
import pandas as pd 
from sqlalchemy import create_engine 

*Устанавливаем параметры:*

In [2]:
db_config = {'user': 'praktikum_student', # имяпользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
             'port': 6432,# портподключения
             'db': 'data-analyst-final-project-db'} # названиебазыданных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],  db_config['pwd'],  db_config['host'],  db_config['port'],  db_config['db'])  

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

<a id='data_import'></a>
### Откроем таблицы с данными и изучим общую информацию:

In [3]:
query = """
  SELECT *
  FROM books;
"""

pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
query = """
  SELECT *
  FROM authors;
"""

pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [5]:
query = """
  SELECT *
  FROM publishers;
"""

pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [6]:
query = """
  SELECT *
  FROM ratings;
"""

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [7]:
query = """
  SELECT *
  FROM reviews;
"""

pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

Здесь всё хорошо. Ты корректно выгрузил данные и посмотрел их содержимое. Вывел информацию  о размерности датасетов. Теперь можем приступать к выполнению заданий
</div>

[В начало](#contents)

<a id='data_analysis'></a>
## Анализ данных:

*Посчитаем, сколько книг вышло после 1 января 2000 года:*

In [8]:
query = """
 SELECT
      COUNT(books.book_id) AS books_count
 FROM
    books
 WHERE 
    books.publication_date :: date >= '2000-01-01' :: date;
"""

pd.io.sql.read_sql(query, con = engine)

books_count
0          821

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Да, теперь всё верно
</div>

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

Не совсем верный ответ, т.к. в задаче речь идет о дате 01.01.2000, а у тебя 01.12.2000
</div>

In [9]:
query = """
  SELECT MAX(publication_date :: date)
  FROM books;
"""

pd.io.sql.read_sql(query, con = engine)

max
0  2020-03-31

In [10]:
print('Количество книг, которые выходят в месяц:', 821 / ((20 * 12) + 3))

Количество книг, которые выходят в месяц: 3.3786008230452675


**Вывод:**

- Если не брать разбивку по издательствам, то в целом книги выпускаются достаточно бодро.

*Для каждой книги посчитаем количество обзоров и среднюю оценку:*

In [11]:
query = """
 SELECT
   ratings.book_id,
   AVG(rating) AS avg_rating,
   subq.reviews_count AS reviews_count
 FROM 
   ratings
 LEFT JOIN
  (
   SELECT
    reviews.book_id,
    COUNT(reviews.review_id) AS reviews_count
   FROM
    reviews
   GROUP BY
    reviews.book_id
  ) AS subq 
 ON 
   ratings.book_id = subq.book_id
 WHERE
   subq.book_id IS NOT NULL
 GROUP BY
    ratings.book_id,
    subq.reviews_count
 ORDER BY
    avg_rating DESC;
"""

pd.io.sql.read_sql(query, con = engine)

book_id  avg_rating  reviews_count
0        321        5.00              2
1        224        5.00              2
2        213        5.00              2
3        518        5.00              2
4        469        5.00              2
..       ...         ...            ...
989      915        2.25              3
990      316        2.00              2
991      202        2.00              3
992      371        2.00              2
993      303        1.50              2

[994 rows x 3 columns]

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Здорово. Это один из самых коварных запросов, как показывает практика, но ты справился успешно. Молодец!)
</div>

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

Ошибка воспроизведения. Рекомендую перед отправкой всегда прогонять работу через Kernel-Restart&Run All
</div>

**Вывод:**

- Средний рейтинг не влияет на количество обзоров, и наоборот.

*Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (исключим количеством страниц из анализа брошюры):*

In [12]:
query = """
 SELECT
   publishers.publisher_id,
   publisher,
   MAX(subq.books_count) AS max_publishes
 FROM 
   publishers
 LEFT JOIN
  (
  SELECT 
    books.publisher_id,
    COUNT(books.book_id) AS books_count
  FROM books
  WHERE books.num_pages > 50
  GROUP BY
    books.publisher_id
  ) AS subq 
 ON 
   publishers.publisher_id = subq.publisher_id
 WHERE
   subq.publisher_id IS NOT NULL
 GROUP BY
    publishers.publisher_id
 ORDER BY
     max_publishes DESC
 LIMIT 
     1;
"""

pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  max_publishes
0           212  Penguin Books             42

**Вывод:**

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

Согласен с результатом. Действительно данное издательство является лидером по выпущенным книгам
</div>

- Мы рассматриваем издательства с не самым большим количеством авторских/издательских прав.

*Определим автора с самой высокой средней оценкой книг (учитывайте только книги с 50 и более оценками):*

In [13]:
query = """
 SELECT
    authors.author AS author,
    AVG(subq.avg_rating) AS avg_author_rating
 FROM
    authors
 INNER JOIN
    books ON books.author_id = authors.author_id
 INNER JOIN
   ( 
    SELECT
      book_id,
      AVG(rating) AS avg_rating,
      COUNT(rating_id) AS ratings_count
    FROM
      ratings
    GROUP BY
      book_id
    ) AS subq 
    ON subq.book_id = books.book_id
    WHERE
      subq.ratings_count > 50
  GROUP BY
    author
  ORDER BY
    avg_author_rating DESC
   LIMIT 
     1;  
   """

pd.io.sql.read_sql(query, con = engine)

author  avg_author_rating
0  J.K. Rowling/Mary GrandPré           4.283844

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Такой вариант имеет право быть. Результат корректный
</div>

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

Запрос почти корректный, ты верно отфильтровал книги по условию задачи, но в рейтинг автора ты записал максимальную среднюю оценку одной из его книг (у некоторых авторов несколько произведений), а надо рассчитать средний рейтинг автора как среднее по оценкам его книг
</div>

**Вывод:**

- Оценка книг производилась среди англоязычных пользователей.

*Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок:*

In [14]:
query = """
 SELECT
   COUNT(reviews.username) AS reviewers_count,
   AVG(subq_2.reviews_count) AS avg_reviews_count
 FROM 
   reviews
 LEFT JOIN 
  (
  SELECT
   reviews.username,
   subq_1.rates_count,
   COUNT(review_id) AS reviews_count
  FROM 
   reviews
  LEFT JOIN
   (
   SELECT 
     ratings.username,
     COUNT(rating_id) AS rates_count
   FROM 
     ratings 
   GROUP BY
     ratings.username
   ) AS subq_1
   ON 
    reviews.username = subq_1.username
   WHERE
    subq_1.username IS NOT NULL AND subq_1.rates_count > 50
  GROUP BY
     reviews.username,
     subq_1.rates_count
   ) AS subq_2
   ON 
    reviews.username = subq_2.username
   WHERE
    subq_2.username IS NOT NULL  
"""

pd.io.sql.read_sql(query, con = engine)

reviewers_count  avg_reviews_count
0              146          24.808219

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

И здесь у тебя к сожалению ошибка. Хоть и результат очень близкий.
    
Но запрос явно не оптимальный. Всё гораздо проще (даже джойнов никаких не надо). Я всегда рекомендую первым делом подзапросом выполнять условие задачи. В данном случае, тебе надо отфильтровать пользователей from ratings и оставить тех кто ставил больше 50 оценок. Так мы сразу отсекаем ненужные данные - это оптимальное решение. Т.е. на выходе подзапроса у нас просто имена пользователей username.

Далее нам надо посчитать сколько сделали они ревью. В reviews есть и username и review_id, просто потом фильтруем пользователей через ```where reviews.username in (*твой подзапрос*)```. Ну и найти общее среднее этих показателей. Попробуй)
</div>

<div class="alert alert-block alert-info"> 
    
**Ответ студента:**
    Разбил на два этапа.
    Код ниже.
</div>

*Посмотрим, как соотносится количество оценок и обзоров:*

In [15]:
query = """
 SELECT
  reviews.username,
  subq_1.ratings_count,
  COUNT (reviews.review_id) AS reviews_count
 FROM
  reviews
 LEFT JOIN 
  (
  SELECT 
    ratings.username,
    COUNT(rating_id) AS ratings_count
   FROM
    ratings
   GROUP BY
    ratings.username
   ) AS subq_1
   ON 
    reviews.username = subq_1.username
  WHERE
    subq_1.username IS NOT NULL AND subq_1.ratings_count > 50 AND reviews.username IN (subq_1.username)
    GROUP BY
     reviews.username,
     subq_1.ratings_count  
"""

pd.io.sql.read_sql(query, con = engine)

username  ratings_count  reviews_count
0  jennifermiller             53             25
1       richard89             55             26
2          xdavis             51             18
3          paul88             56             22
4      martinadam             56             27
5     sfitzgerald             55             28

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера v2:</b>

Ок. Моё предложение тебя не устроило, тогда пожалуйста объясни - почему когда я беру среднее значение из твоего первого этапа у меня получается:
</div>

<div class="alert alert-block alert-info"> 
    
**Ответ студента 2:**
 Никогда бы не подумал, что так можно было бы сделать (при мощи .mean).
 Тем не менее, как-то повлиять на работу функции AVG я не могу.
 Возможно ли что-то изменить в коде вычисления среднего черех SQL, чтобы получалось то же значения, учитывая, что составление таблицы с исходными данными полностью совпадает?
 Можно ли считать, что данная работа зачтена с условием того, что я узнал новый способ вычисления реднего значения, полученного через запрос SQL, но не средствами SQL?
 Что могло повлиять на работу AVG функции?
</div>

<div class="alert alert-block alert-warning">
<b>Комментарии от ревьюера:</b>

"новый способ вычисления среднего значения, полученного через запрос SQL" - это не совсем так. С помощью данного метода ```pd.io.sql.read_sql(query, con = engine)``` ты результат запроса закидываешь в датафрейм объект, а к нему можно применять различные агрегирующие функции)
    
"составление таблицы с исходными данными полностью совпадает" -  да, ты верно отфильтровал пользователей и рассчитал количество ревью которое они написали
    
"Что могло повлиять на работу AVG функции?" - давай посмотрим результат последнего этапа без AVG
</div>

In [22]:
# КОД РЕВЬЮЕРА

query = """
 SELECT
  *
 FROM 
   reviews   
LEFT JOIN 
 (
 SELECT
  reviews.username,
  subq_1.ratings_count,
  COUNT (reviews.review_id) AS reviews_count
 FROM
  reviews
 LEFT JOIN 
  (
  SELECT 
    ratings.username,
    COUNT(rating_id) AS ratings_count
   FROM
    ratings
   GROUP BY
    ratings.username
   ) AS subq_1
   ON 
    reviews.username = subq_1.username
  WHERE
    subq_1.username IS NOT NULL AND subq_1.ratings_count > 50 AND reviews.username IN (subq_1.username)
    GROUP BY
     reviews.username,
     subq_1.ratings_count
  ) AS subq_2
   ON 
    reviews.username = subq_2.username
   WHERE
    subq_2.username IS NOT NULL;
    
"""

pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0         2750      985  jennifermiller   
1         2690      962  jennifermiller   
2         2438      872  jennifermiller   
3         2121      763  jennifermiller   
4         2096      753  jennifermiller   
..         ...      ...             ...   
141        180       63     sfitzgerald   
142        169       59     sfitzgerald   
143        150       52     sfitzgerald   
144         79       29     sfitzgerald   
145         16        6     sfitzgerald   

                                                  text        username  \
0    Prove instead wrong town. Police everybody hav...  jennifermiller   
1    Total last same. Enter person trip onto man th...  jennifermiller   
2    Front gas away receive. Say read stuff forward...  jennifermiller   
3    All need anyone build win area personal. Great...  jennifermiller   
4    Appear give matter wrong. Individual informati...  jennifermiller   
..                                                 ...             ...   
141  Mission your goal. Above budget material side ...     sfitzgerald   
142  Stop find moment role point possible accept. W...     sfitzgerald   
143  Mouth provide oil agency occur wear year air. ...     sfitzgerald   
144  Billion own individual general lay any scene. ...     sfitzgerald   
145  Miss effort gun strong surface civil. Resource...     sfitzgerald   

     ratings_count  reviews_count  
0               53             25  
1               53             25  
2               53             25  
3               53             25  
4               53             25  
..             ...            ...  
141             55             28  
142             55             28  
143             55             28  
144             55             28  
145             55             28  

[146 rows x 7 columns]

<div class="alert alert-block alert-warning">
<b>Комментарии от ревьюера v2:</b>

Как видно, ты к таблице reviews добавил reviews_count\ratings_count и только потом взял среднее. Надеюсь видно, что это было плохой идеей)
    
Я хотел подтолкнуть тебя к более простому и оптимальному варианту. Эту задачу можно было решить без джойнов. 
    
1. Первым шагом мы фильтруем пользователей (```query_rev_01```)
    
2. В reviews есть и username и review_id, просто потом фильтруем пользователей через ```where reviews.username in (*твой подзапрос*)``` (```query_rev_02```)
    
3. Находим среднее значение. Для удобство восприятия\чтения запроса использую подход с with as (```query_rev_03```)
</div>

In [41]:
# КОД РЕВЬЮЕРА

query_rev_01 = (
'''
    select username
    from ratings
    group by username
    having count(rating_id) > 50
    
''')

pd.io.sql.read_sql(query_rev_01, con = engine)

username
0     sfitzgerald
1  jennifermiller
2          xdavis
3          paul88
4      martinadam
5       richard89

In [42]:
# КОД РЕВЬЮЕРА

query_rev_02 = (
'''
    select reviews.username, count(review_id) as cnt_rev
    from reviews
    where reviews.username in ( select username
                        from ratings
                        group by username
                        having count(rating_id) > 50)
    group by reviews.username
    
''')

pd.io.sql.read_sql(query_rev_02, con = engine)

username  cnt_rev
0     sfitzgerald       28
1  jennifermiller       25
2          xdavis       18
3          paul88       22
4      martinadam       27
5       richard89       26

In [43]:
# КОД РЕВЬЮЕРА

query_rev_03 = (
'''
with table_full as 
    ( 
      select reviews.username, count(review_id) as cnt_rev
      from reviews
      where reviews.username in ( select username
                                  from ratings
                                  group by username
                                  having count(rating_id) > 50 )
      group by reviews.username
    )
    
select avg(cnt_rev)
from table_full
    
''')

pd.io.sql.read_sql(query_rev_03, con = engine)

avg
0  24.333333

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера v2:</b>

А у тебя 24.808219
</div>

*Найдём среднее среди количества обзоров:*

In [19]:
query = """
 SELECT
  AVG(subq_2.reviews_count) AS avg_reviews_count
 FROM 
   reviews   
LEFT JOIN 
 (
 SELECT
  reviews.username,
  subq_1.ratings_count,
  COUNT (reviews.review_id) AS reviews_count
 FROM
  reviews
 LEFT JOIN 
  (
  SELECT 
    ratings.username,
    COUNT(rating_id) AS ratings_count
   FROM
    ratings
   GROUP BY
    ratings.username
   ) AS subq_1
   ON 
    reviews.username = subq_1.username
  WHERE
    subq_1.username IS NOT NULL AND subq_1.ratings_count > 50 AND reviews.username IN (subq_1.username)
    GROUP BY
     reviews.username,
     subq_1.ratings_count
  ) AS subq_2
   ON 
    reviews.username = subq_2.username
   WHERE
    subq_2.username IS NOT NULL;
    
"""

pd.io.sql.read_sql(query, con = engine)

avg_reviews_count
0          24.808219

**Вывод:**

- Особо активных читателей не так много, но они очень продуктивны в плане обзоров. Возможно, сказывается начитанность, выливающаяся в умение писать тексты.

[В начало раздела](#data_analysis)

[В начало](#contents)

<a id='general_conclusion'></a>
## Общий вывод

- Количество книг, которые выходят в месяц за рубежом в стране, где проводилось исследование: 3.3786008230452675
- Если не брать разбивку по издательствам, то в целом книги выпускаются достаточно бодро.
- Средний рейтинг не влияет на количество обзоров, и наоборот.
- Мы рассматриваем издательства с не самым большим количеством авторских/издательских прав.
- Оценка книг производилась среди англоязычных пользователей.
- Особо активных читателей не так много, но они очень продуктивны в плане обзоров. Возможно, сказывается начитанность, выливающаяся в умение писать тексты.

[В начало](#contents)